src/select.c

```diff
@@ -263,13 +258,13 @@ int W32_CALL select_s (int nfds, fd_set *readfds, fd_set *writefds,
       total_ex += exc_cnt;
 #endif
 
-      /* Safe to run other "tasks" now.
-       */
-      _sock_crit_stop();
       SOCK_DBUG_FLUSH();
 
     } /* end of for loop; all sockets checked at least once */
 
+    /* Safe to run other "tasks" now.
+     */
+    _sock_crit_stop();
 
     /* WATT_YIELD() sometimes hangs for approx 250msec under Win-XP.
      * Don't yield for "small" timeouts.
```

Comment:
BTW, there is this bit in `select_s()`.  Do we want the same logic for `poll()` too?  Or maybe remove it from `select_s()` (don't know how relevant WinXP still is).

Problem here is, if someone did `sock_yield(NULL, my_yield)`, that will never be called either.

Reply:
It should yield if `ret_count == 0` in the hope that some events happen in the mean time.

>  (don't know how relevant WinXP still is).

Not sure if it's an issue with Win-10. Some test is needed to verify.

> Problem here is, if someone did sock_yield(NULL, my_yield), that will never be called either.

That `my_yield` won't be called? I think it should be called indirectly as:
```c
      if (socket->tcp_sock->usr_yield)
         (*socket->tcp_sock->usr_yield) ();
```
in lots of places.

Reply:
> It should yield if `ret_count == 0` in the hope that some events happen in the mean time.

It does, but `select_s()` skips that if the timeout is less than 500ms.  Seems to be the only place where that is done.

Looking at `WATT_YIELD` again, I see now it expands to `os_yield()` (eg. `int 2f, ax=1680`) directly.  Huh.  I thought it called `system_yield`.  Maybe it should?

Reply:
>  Maybe it should?

I don't see a problem here. What is it?

Reply:
There are two, basically:
* `WATT_YIELD()` doesn't call the user-provided yield function
* `select_s()` doesnt call `WATT_YIELD()` if timeout is less than 500ms

I think if the user has set a custom global yield function, that should take precedence over the OS one.  
Then for WinXP maybe some other workaround can be found (eg. only yield once every X ms).

And again, thanks for merging!

Reply:
Do you happen to remember if this comment in `select.c` applies to a native winxp program, or a dos program running under winxp?  I'm guessing it's only the latter?

Reply:
>  I'm guessing it's only the latter?

AFAICR, yes. I even ran djgpp/Watt-32 programs under Win-95 (or was it Win-98?). 
I think Win-XP made it worse for djgpp; hence  that comment. But again, I'm rusty on these details after such a long time.
And now after 2 pints of Tuborgs.

Reply:
I can sort-of replicate this on windows 2000, but I'm not sure if it's the same thing.  Using a djgpp test program, doing the following in a loop:

```c
t0 = set_timeout(0);
__dpmi_yield();
t = set_timeout(0) - t0;
```

I'm regularly seeing `t` values around 1000ms.  But it doesn't feel like it's stalling for a whole second.  And it often jumps backwards too (`t < -900`)...  Maybe the rdtsc code isn't very accurate on Windows?  The high `t` values also completely disappear after using `init_timer_isr()` (which disables rdtsc stuff).